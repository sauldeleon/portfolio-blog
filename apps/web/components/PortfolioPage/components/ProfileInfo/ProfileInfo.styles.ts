import { styled } from 'styled-components'

import { Body, Label } from '@sdlgr/typography'

export const StyledSummary = styled(Body)`
  font-size: 20px;

  ${({ theme }) => theme.media.up.lg} {
    font-size: 30px;
  }
`

export const StyledLabel = styled(Label)`
  margin-top: 40px;
`

export const StyledSkillGroup = styled.div`
  margin-top: 20px;
`

export const StyledSkillTitle = styled(Body)`
  font-weight: 500;
  display: flex;
  line-height: 20px;
  margin-bottom: 10px;

  svg {
    margin-right: 10px;
  }
`

export const StyledList = styled.ul`
  padding-left: 0px;

  ${({ theme }) => theme.media.up.lg} {
    padding-left: 25px;
  }
`

export const StyledListItem = styled.li`
  font-size: 16px;
  text-indent: -20px;
  padding-left: 20px;

  &:before {
    content: 'Â·';
    font-size: 30px;
    vertical-align: middle;
    line-height: 20px;
  }
`

export const StyledTechnology = styled.b`
  ${({ theme }) => theme.typography.body.M}
  font-size: 16px;
  font-weight: bold;
`

export const StyledItalic = styled.i`
  ${({ theme }) => theme.typography.body.M}
  font-size: 16px;
`
