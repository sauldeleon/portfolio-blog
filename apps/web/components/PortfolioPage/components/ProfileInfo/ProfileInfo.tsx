import { ParseKeys } from 'i18next'
import { Trans } from 'react-i18next'

import {
  BackendIcon,
  DatabaseIcon,
  FileIcon,
  FrontendIcon,
  MobileIcon,
} from '@sdlgr/assets'

import { useClientTranslation } from '@web/i18n/client'

import {
  StyledItalic,
  StyledLabel,
  StyledList,
  StyledListItem,
  StyledSkillGroup,
  StyledSkillTitle,
  StyledSummary,
  StyledTechnology,
} from './ProfileInfo.styles'

const iconMap: Record<string, React.ReactElement> = {
  frontEnd: <FrontendIcon width={20} height={20} />,
  backEnd: <BackendIcon width={20} height={20} />,
  mobile: <MobileIcon width={20} height={20} />,
  database: <DatabaseIcon width={20} height={20} />,
  other: <FileIcon width={20} height={20} />,
}

export function ProfileInfo() {
  const { t } = useClientTranslation('portfolioPage')

  const areas = t('items.profile.skillAreas.areas', {
    returnObjects: true,
  })

  return (
    <>
      <StyledSummary $level="XL">{t('items.profile.summary')}</StyledSummary>
      <StyledLabel>{t('items.profile.skillAreas.title')}</StyledLabel>
      {areas.map(({ icon, title, skills }) => (
        <StyledSkillGroup key={icon}>
          <StyledSkillTitle>
            {iconMap[icon]}
            {title}
          </StyledSkillTitle>
          <StyledList>
            {skills.map((skill) => (
              <StyledListItem key={skill}>
                <Trans
                  t={t}
                  i18nKey={skill as ParseKeys<'portfolioPage'>}
                  components={{
                    bold: <StyledTechnology />,
                    italic: <StyledItalic />,
                  }}
                />
              </StyledListItem>
            ))}
          </StyledList>
        </StyledSkillGroup>
      ))}
    </>
  )
}
