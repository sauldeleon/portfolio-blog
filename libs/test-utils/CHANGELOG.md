# Changelog Test utils

## [0.2.3](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.2...test-utils-0.2.3) (2024-01-08)

## [0.2.2](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.1...test-utils-0.2.2) (2023-09-26)

## [0.2.1](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.0...test-utils-0.2.1) (2023-09-14)

# [0.2.0](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.1.0...test-utils-0.2.0) (2023-09-05)

### Bug Fixes

- reduced duplicated lines ([f51168a](https://github.com/sauldeleon/portfolio-blog/commit/f51168aefe025a5a65ffdb2dc22f7a16b863406e))
- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))

# 0.1.0 (2023-09-02)

### Bug Fixes

- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- add contact initial page ([fea862b](https://github.com/sauldeleon/portfolio-blog/commit/fea862b3507268013149050ae3f05fa517e7b3d9))
- add homepage ([60523aa](https://github.com/sauldeleon/portfolio-blog/commit/60523aa31ec088d318745ef0c757b4cbe656f0d4))
- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))
- improved a11y and styles ([873f656](https://github.com/sauldeleon/portfolio-blog/commit/873f656587f50df8e29ad526fdc68fea0b2c9784))
- toggle language button ([afd97e9](https://github.com/sauldeleon/portfolio-blog/commit/afd97e95beac0f70ed643c753e71db2ca014054b))
