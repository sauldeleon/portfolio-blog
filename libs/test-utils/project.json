{
  "name": "test-utils",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/test-utils/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "version": {
      "executor": "@jscutlery/semver:version",
      "options": {
        "changelogHeader": "# Changelog Test utils \n",
        "baseBranch": "main",
        "commitMessageFormat": "chore(${projectName}): release: v${version}",
        "push": false
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/test-utils/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/test-utils/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  }
}
