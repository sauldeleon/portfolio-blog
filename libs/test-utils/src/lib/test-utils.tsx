import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { Matcher } from '@testing-library/dom/types/matches'
import {
  RenderOptions,
  RenderResult,
  queryAllByAttribute as _queryAllByAttribute,
  queryByAttribute as _queryByAttribute,
  act,
  render,
  screen,
} from '@testing-library/react'
import { RouterContext } from 'next/dist/shared/lib/router-context'
import Router, { NextRouter } from 'next/router'
import { ReactNode, useEffect, useState } from 'react'
import { ThemeProvider } from 'styled-components'

import { LanguageContextProvider } from '@sdlgr/i18n-config'
import { mainTheme } from '@sdlgr/main-theme'

export type MatcherCustomFnParams = {
  content: string
  element: Element | null
}

/* istanbul ignore next */
export const renderWithTheme = (children: ReactNode): RenderResult => {
  return render(<ThemeProvider theme={mainTheme}>{children}</ThemeProvider>)
}

/* istanbul ignore next */
export function RenderProviders({ children }: { children: ReactNode }) {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  return (
    <QueryClientProvider client={queryClient}>
      <LanguageContextProvider value={{ language: 'en' }}>
        <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
      </LanguageContextProvider>
    </QueryClientProvider>
  )
}

/* istanbul ignore next */
export const renderApp = (
  children: ReactNode,
  options?: Omit<RenderOptions, 'queries'>
) => {
  return render(<RenderProviders>{children}</RenderProviders>, options)
}

/**
 * Generic function to get element by attribute and value
 * @example queryByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function queryByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example queryByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function queryAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Generic function to get element by attribute and value
 * @example getByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function getByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example getByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function getAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Shortcut version for queryByAttribute, but specific for id
 * @param {HTMLElement} container
 * @param {string} id
 * @return {any}
 */
export function getById(container: HTMLElement, id: string) {
  return queryByAttribute('id')(container, id) as Element
}

/**
 * Shortcut version for queryByAttribute, but specific for class name
 * @param {HTMLElement} container
 * @param {string} className
 * @return {any}
 */
export function getByClassName(container: HTMLElement, className: string) {
  return queryByAttribute('class')(container, className) as Element
}

function matcherCustomFn(
  tag: string,
  matcher: Matcher,
  { element, content }: MatcherCustomFnParams
) {
  const isTag = element?.tagName.toUpperCase() === tag.toUpperCase()
  let matching = false
  if (isTag) {
    matching = content === matcher
    /* istanbul ignore else*/
    if (typeof matcher === 'function') {
      matching = matcher(content, element)
    } else if (matcher instanceof RegExp) {
      matching = matcher.test(content)
    }
  }
  return matching && isTag
}

/**
 * Get 1 element by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getByTagAndText(tag: string, matcher: Matcher) {
  return screen.getByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Get multiple elements by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getAllByTagAndText(tag: string, matcher: Matcher) {
  return screen.getAllByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Mock Math.random() to return a specific value
 * @param {number} value
 */
export function mockRandom(value = 0.1234) {
  const mockMath = Object.create(global.Math)
  mockMath.random = () => value
  global.Math = mockMath
}

/**
 * Hook to use in test to ensure "last" render is called
 */
export function useWithRefreshInTest() {
  const [withRefreshInTest, setWithRefreshInTest] = useState(false)

  useEffect(() => {
    // Change state to "rerender" the component
    setWithRefreshInTest(true)
  }, [])

  return withRefreshInTest
}

export { render } from '@testing-library/react'

/* istanbul ignore next */
export const awaitAllAsync = async () =>
  act(() => new Promise((resolve) => setTimeout(resolve, 0)))

/* istanbul ignore next */
export function renderWithNextRouter(
  children: ReactNode,
  options?: Partial<NextRouter>
) {
  return renderWithTheme(
    <RouterContext.Provider value={{ ...Router, ...options }}>
      {children}
    </RouterContext.Provider>
  )
}
