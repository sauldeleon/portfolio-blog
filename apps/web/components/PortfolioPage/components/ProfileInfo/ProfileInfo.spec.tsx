import { screen } from '@testing-library/react'

import { renderApp } from '@sdlgr/test-utils'

import { ProfileInfo } from './ProfileInfo'

describe('ProfileInfo', () => {
  it('should render successfully', async () => {
    const { baseElement } = renderApp(<ProfileInfo />)
    expect(await screen.findByText('Skills')).toBeInTheDocument()

    expect(baseElement).toMatchSnapshot()
  })
})
