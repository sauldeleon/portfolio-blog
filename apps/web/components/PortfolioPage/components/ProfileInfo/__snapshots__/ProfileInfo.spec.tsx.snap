// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileInfo should render successfully 1`] = `
.c0 {
  font-family: fontFamily;
  font-weight: 500;
  color: #FBFBFB;
  font-size: 1.5rem;
  line-height: normal;
  white-space: pre-line;
}

.c0 ul,
.c0 ol {
  margin-left: 40px;
  margin-bottom: 10px;
}

.c0 ul {
  list-style: outside disc;
}

.c0 ol {
  list-style: decimal;
}

.c0 li {
  padding-left: 10px;
}

.c5 {
  font-family: fontFamily;
  font-weight: 400;
  color: #FBFBFB;
  font-size: 0.9rem;
  line-height: normal;
  white-space: pre-line;
}

.c5 ul,
.c5 ol {
  margin-left: 40px;
  margin-bottom: 10px;
}

.c5 ul {
  list-style: outside disc;
}

.c5 ol {
  list-style: decimal;
}

.c5 li {
  padding-left: 10px;
}

.c2 {
  font-family: fontFamily;
  font-weight: 500;
  color: #FBFBFB;
  font-size: 1rem;
  line-height: normal;
  text-transform: uppercase;
}

.c1 {
  font-size: 20px;
}

.c3 {
  margin-top: 40px;
}

.c4 {
  margin-top: 20px;
}

.c6 {
  font-weight: 500;
  display: flex;
  line-height: 20px;
  margin-bottom: 10px;
}

.c6 svg {
  margin-right: 10px;
}

.c7 {
  padding-left: 0px;
}

.c8 {
  font-size: 16px;
  text-indent: -20px;
  padding-left: 20px;
}

.c8:before {
  content: 'Â·';
  font-size: 30px;
  vertical-align: middle;
  line-height: 20px;
}

.c9 {
  font-family: fontFamily;
  font-weight: 400;
  color: #FBFBFB;
  font-size: 0.9rem;
  line-height: normal;
  font-size: 16px;
  font-weight: bold;
}

.c10 {
  font-family: fontFamily;
  font-weight: 400;
  color: #FBFBFB;
  font-size: 0.9rem;
  line-height: normal;
  font-size: 16px;
}

@media (width >= 1024px) {
  .c1 {
    font-size: 30px;
  }
}

@media (width >= 1024px) {
  .c7 {
    padding-left: 25px;
  }
}

<body>
  <div>
    <p
      class="c0 c1"
    >
      Experienced Software Engineer skilled in translating code into polished and efficient digital solutions. With a meticulous attention to detail, I specialize in developing software that not only fulfills functional requirements but also results into a smooth and user-friendly experience.
    </p>
    <div
      class="c2 c3"
    >
      Skills
    </div>
    <div
      class="c4"
    >
      <p
        class="c5 c6"
      >
        <svg
          height="20"
          width="20"
        >
          frontEnd.svg
        </svg>
        Front End
      </p>
      <ul
        class="c7"
      >
        <li
          class="c8"
        >
          SPA using 
          <b
            class="c9"
          >
            React 18
          </b>
           and 
          <b
            class="c9"
          >
            yarn
          </b>
           as the main package manager.
        </li>
        <li
          class="c8"
        >
          SSR using 
          <b
            class="c9"
          >
            Next.JS 14
          </b>
           with experience in both the new layout system and the legacy system.
        </li>
        <li
          class="c8"
        >
          Testing: Primarily using 
          <b
            class="c9"
          >
            Jest
          </b>
           for unit testing. 
          <b
            class="c9"
          >
            Cypress
          </b>
           for e2e testing
        </li>
        <li
          class="c8"
        >
          Usage of 
          Storybook
           for documenting components, design systems and applications
        </li>
        <li
          class="c8"
        >
          Use of CSS-in-JS libraries such as 
          <b
            class="c9"
          >
            Emotion
          </b>
           or 
          <b
            class="c9"
          >
            Styled-Components
          </b>
          .
        </li>
        <li
          class="c8"
        >
          Preference for projects in 
          <b
            class="c9"
          >
            Typescript
          </b>
          .
        </li>
        <li
          class="c8"
        >
          Experience with 
          <b
            class="c9"
          >
            monorepos
          </b>
           (lerna, NX, yarn workspaces...).
        </li>
      </ul>
    </div>
    <div
      class="c4"
    >
      <p
        class="c5 c6"
      >
        <svg
          height="20"
          width="20"
        >
          backEnd.svg
        </svg>
        Back End
      </p>
      <ul
        class="c7"
      >
        <li
          class="c8"
        >
          <b
            class="c9"
          >
            REST
          </b>
           applications in node.js using 
          <b
            class="c9"
          >
            express
          </b>
           and 
          <b
            class="c9"
          >
            bottle.js
          </b>
           for dependency injection. 
          <b
            class="c9"
          >
            Sequelize
          </b>
           and 
          <b
            class="c9"
          >
            mongoose
          </b>
           for integration with different databases.
        </li>
        <li
          class="c8"
        >
          Java development experience with 
          <b
            class="c9"
          >
            Spring Framework
          </b>
           (Boot, Core, Security, MVC). Use of 
          <b
            class="c9"
          >
            Hibernate
          </b>
           for integration with Oracle, MySQL, MongoDB... Experience with Apache Tiles and JMX, integrations with 
          <b
            class="c9"
          >
            Kafka
          </b>
          .
        </li>
        <li
          class="c8"
        >
          Development of 
          <b
            class="c9"
          >
            Perl
          </b>
           applications using 
          <b
            class="c9"
          >
            Catalyst
          </b>
           as the main MVC framework.
        </li>
      </ul>
    </div>
    <div
      class="c4"
    >
      <p
        class="c5 c6"
      >
        <svg
          height="20"
          width="20"
        >
          mobile.svg
        </svg>
        Mobile Apps
      </p>
      <ul
        class="c7"
      >
        <li
          class="c8"
        >
          Experience with 
          <b
            class="c9"
          >
            EXPO
          </b>
           over 
          <b
            class="c9"
          >
            React Native
          </b>
           for developing native applications.
        </li>
        <li
          class="c8"
        >
          Basic knowledge of 
          <b
            class="c9"
          >
            IONIC
          </b>
           and 
          <b
            class="c9"
          >
            Cordova
          </b>
           for web-in-app development.
        </li>
      </ul>
    </div>
    <div
      class="c4"
    >
      <p
        class="c5 c6"
      >
        <svg
          height="20"
          width="20"
        >
          database.svg
        </svg>
        Databases
      </p>
      <ul
        class="c7"
      >
        <li
          class="c8"
        >
          Course 
          <i
            class="c10"
          >
            M101J, MongoDB for Java Developers
          </i>
           offered by MongoDB Inc. Experience with integrating MongoDB with Java, Perl, and Node.js.
        </li>
        <li
          class="c8"
        >
          Experience with Oracle, PostgreSQL (including the 
          <b
            class="c9"
          >
            PostGIS
          </b>
           module), and MySQL in Java, Perl, and Node.js.
        </li>
      </ul>
    </div>
    <div
      class="c4"
    >
      <p
        class="c5 c6"
      >
        <svg
          height="20"
          width="20"
        >
          file.svg
        </svg>
        Others
      </p>
      <ul
        class="c7"
      >
        <li
          class="c8"
        >
          Working environment on Windows, Unix, or Mac.
        </li>
        <li
          class="c8"
        >
          Experience with 
          <b
            class="c9"
          >
            Gitlab Pipelines
          </b>
          , 
          <b
            class="c9"
          >
            Github Actions
          </b>
          , and 
          <b
            class="c9"
          >
            Jenkins
          </b>
           for CI/CD.
        </li>
      </ul>
    </div>
  </div>
</body>
`;
